tweet_recommendations/our_method.py

```diff
@@ -0,0 +1,286 @@
+        :return: nx.Graph of tweets & hashtags.
+        """
+
+        G = nx.Graph()
+
+        def add_row_to_graph(row, graph, embedding_name):
+            if len(row.hashtags) > 0:
+                graph.add_node(row.id)
+                graph.node[row.id]["node_type"] = "tweet"
+                graph.node[row.id]["retweets"] = row.retweet_count
+                graph.node[row.id][embedding_name] = row[embedding_name]
+
+                for hashtag in row.hashtags:
+                    graph.add_node(hashtag)
+                    graph.node[hashtag]["node_type"] = "hashtag"
+                    graph.add_edge(row.id, hashtag)
+
+        if self.verbose:
+            print("Building graph ...")
+            tqdm.tqdm.pandas()
+            tweets.progress_apply(lambda r:
+                                  add_row_to_graph(r, G, self.embedding_name),
+                                  axis=1)
+        else:
+            tweets.apply(lambda r: add_row_to_graph(r, G, self.embedding_name),
+                         axis=1)
+        return G
+
+    def _calculate_hashtag_embeddings(self, G: nx.Graph):
+
+        for node in tqdm.tqdm(G.nodes, total=len(G.nodes),
+                              disable=not self.verbose):
+            if G.nodes[node]["node_type"] == "hashtag":
+                tweets = list(G.neighbors(node))
+                embeddings = np.asarray([G.nodes[tweet][self.embedding_name]
+                                         for tweet in tweets])
+                G.nodes[node][self.embedding_name] = embeddings.mean(axis=0)
+
+        return G
+
+    def _calculate_pagerank(self, G: nx.Graph):
+        graph_pagerank = nx.pagerank(G)
+        nx.set_node_attributes(G, graph_pagerank, "pagerank")
+        return G
+
+    def _calculate_hashtag_mean_retweets(self, G: nx.Graph):
+        for node in tqdm.tqdm(G.nodes, total=len(G.nodes),
+                              disable=not self.verbose):
+            if G.nodes[node]["node_type"] == "hashtag":
+                tweets = G.neighbors(node)
+                retweets_counts = np.asarray([G.nodes[tweet]['retweets']
+                                              for tweet in tweets])
+                G.nodes[node]['mean_retweets'] = retweets_counts.mean(axis=0)
+        return G
+
+    def _calculate_edge_weights(self, G: nx.Graph,
+                                distance_name: str = 'distance',
+                                similarity_name: str = 'similarity'):
+        for (node_from,
+             node_to, edge_features) in tqdm.tqdm(G.edges(data=True),
+                                                  total=len(G.edges),
+                                                  disable=not self.verbose):
+            emb_from = G.nodes[node_from][self.embedding_name]
+            emb_to = G.nodes[node_to][self.embedding_name]
+
+            # clipping to avoid numerical errors
+            distance = np.clip(scipy.spatial.distance.cosine(emb_from, emb_to),

```

Comment:
1. `cosine` zwraca wartości z przedziału [-1, 1], a Ty na siłę klipujesz do [0, 1].
2. To liczenie zaimplementowałeś już w `embedding_similarity`, ale tego nie wykorzystujesz tutaj nawet (jakbyś wykorzystał, to by nie było tego błędu 1.  :stuck_out_tongue: )

Reply:
Ups, racja, wykorzystam to co w `embedding_similarity`. 
Błędu by nie było, bo tam też clippuję :P
Namierzyłem takie dwa wektory dla których rzeczywiście wychodzi ponad 1. Jak liczyłem ręcznie z wzorku to to samo i ciężko stwierdzic dlaczego dokładnie...

Reply:
Ah, czyli `cosine` zwraca od 0 (wektory takie same) do 2 (wektory kompletnie różne), a nie od -1 do 1 ... czyli w obu przypadkach jest niepoprawnie clipowane